### ISSUE TITLE
Avoid SC1087 in regex context?

### ISSUE THREAD

**Issue by Tamaranch** [koalaman/shellcheck#1772]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC1087
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
a='a'
[[ $1 =~ $a[0-9] ]]

```

#### Here's what shellcheck currently says:
Line 4:
[[ $1 =~ $a[0-9] ]]
         ^-- SC1087: Use braces when expanding arrays, e.g. ${array[idx]} (or ${var}[.. to quiet).

#### Here's what I wanted or expected to see:
Nothing: in regex context, the meaning of this kind of expression is clear. The wiki page advises to "use ${var}[idx] to prevent this warning", because "this does not change how the script works, but clarifies your intent to ShellCheck as well as other programmers", but my feeling is that it is uselessly heavy in this context (among others).

**Comment by bugfood** [comment]
From testing, it seems that quoting the variable itself will quiet this error. In your example:
```
[[ $1 =~ "$a"[0-9] ]]
```
I think this is an ok workaround as long as the variable part should be matched as a string without respect to regex metacharacters.

As of 5.1.8, the bash manual says:
```
Any part of the pattern may be quoted to force the quoted portion to be
matched as a string.  Bracket expressions in regular expressions must
be treated carefully, since  normal  quoting characters  lose their
meanings between brackets.  If the pattern is stored in a shell variable,
quoting the variable expansion forces the entire pattern to be matched
as a string.
```

If the variable part needs to be interpreted as a regex, then an alternative workaround would be to compose the regex into another variable prior to matching:
```
b="$a"'[0-9]'
[[ $1 =~ $b ]]
```

**Comment by koalaman** [comment]
It's somewhat clear in this case, but you can imagine `[[ $1 =~ $regex[email] ]]` instead, which is structurally the same. This warning is one of those primarily there to help people coming from other languages where `$arr[x]` is fine, and is not nearly as helpful on an ongoing basis for people who already know shell well. Please feel entirely free to banish the warning forever with `echo disable=SC1087 >> ~/.shellcheckrc`, since you're already aware of it.
